src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -0,0 +1,275 @@
+
+        '''
+        self.validate(node)
+
+        array_reference = node.parent
+        array_index = array_reference.children.index(node)
+        assignment = array_reference.parent
+        parent = assignment.parent
+        symbol_table = node.scope.symbol_table
+
+        # See if there is any configuration information for this array index
+        loop_type_order = Config.get().api_conf("nemo").get_index_order()
+        loop_type_data = Config.get().api_conf("nemo").get_loop_type_data()
+        try:
+            loop_type = loop_type_order[array_index]
+            loop_type_info = loop_type_data[loop_type]
+            lower_bound_info = loop_type_info['start']
+            upper_bound_info = loop_type_info['stop']
+            loop_variable_name = loop_type_info['var']
+        except IndexError:
+            lower_bound_info = None
+            upper_bound_info = None
+            loop_variable_name = symbol_table.new_symbol_name("idx")
+
+        # Lower bound
+        if not array_reference.is_lower_bound(array_index):
+            # The range specifies a lower bound so use it
+            lower_bound = node.start
+        elif lower_bound_info:
+            # The config metadata specifies a lower bound so use it
+            try:
+                _ = int(lower_bound_info)
+                lower_bound = Literal(lower_bound_info, INTEGER_TYPE)
+            except ValueError:
+                lower_bound = Reference(symbol_table.lookup(lower_bound_info))
+        else:
+            # The lower bound is not set or specified so use the
+            # LBOUND() intrinsic
+            lower_bound = node.start
+
+        # Upper bound
+        if not array_reference.is_upper_bound(array_index):
+            # The range specifies an upper bound so use it
+            upper_bound = node.start
+        elif upper_bound_info:
+            # The config metadata specifies an upper bound so use it
+            try:
+                _ = int(upper_bound_info)
+                upper_bound = Literal(upper_bound_info, INTEGER_TYPE)
+            except ValueError:
+                upper_bound = Reference(symbol_table.lookup(upper_bound_info))
+        else:
+            # The upper bound is not set or specified so use the
+            # UBOUND() intrinsic
+            upper_bound = node.stop
+
+        # Just use the specified step value
+        step = node.step
+
+        # Look up the loop variable in the symbol table. If it does
+        # not exist then create it.
+        try:
+            loop_variable_symbol = symbol_table.lookup(loop_variable_name)
+        except KeyError:
+            # Add loop variable as it does not already exist
+            loop_variable_symbol = DataSymbol(loop_variable_name, INTEGER_TYPE)
+            symbol_table.add(loop_variable_symbol)
+
+        # Replace the loop_idx array dimension with the loop variable.
+        for array in assignment.walk(ArrayReference):
```

Comment:
I think this needs more care. e.g. NEMO contains code like:

    my_array(:,:) = input(:,:) + some_fn(input2(:,:,:))

Reply:
Yes, that was sloppy. I think the transformation is now robust. It should now only making changes where it is valid to do so. I've supported this by only modifying a range if all parents of the array reference that contains this range node are operators, up to the assignment node. The one additional case I've ruled out is array valued operators and a transformation error is now raised if one of these (currently only matmul) is found. The reason for this is that we need to know the dimension of the array returned by the operator to determine how many dimensions (if any) could be transformed to explicit loops.